### PR TITLE
Parse (more) DPH for GA

### LINKED ISSUE #181: parse department of public health from Georgia
[learn our pipeline: parse](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#parse)

Blocked by #180 

Parse existing data into [`.ndjson`](http://ndjson.org/).

Read all files in the directory passed as the second argument (`sys.argv[2]`), convert them to `.ndjson`, and then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the parse stage and how to get set up for development!

### Tips

1. Fetch data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch ga/dph
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest parse ga/dph
    ```

### Example
[Fetched files for `md/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/md/arcgis/raw) are converted to [parsed files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/md/arcgis/parsed)

### PR DESCRIPTION
**Pull request by Graham42**
# Parse (more) DPH for GA

| Key Details |
|-|
| Resolves #181 |
State: GA |
Site: DPH |

## Notes
The first 2 commits here should be ignored because this branch starts where https://github.com/CAVaccineInventory/vaccine-feed-ingest/pull/228 leaves off. When that get's merged I'll rebase this on main to clear them out.

## Data sample

```json
{"Location Name": "1st Care Management", "County": "Fulton", "Address": "1120 Hope Rd\nSuite 210\nSandy Springs, GA 30350", "address-parts": {"address-line1": "1120 Hope Rd", "address-line2": "Suite 210", "locality": "Sandy Springs", "administrative-area": "GA", "postal-code": "30350", "country": ""}, "phone-numbers": [], "contact-links": [{"href": "https://1stcaremanagement.com/", "label": "COVID-19 Vaccine Appointments"}], "node-id": "node/42871", "last-updated": "1611102629"}
{"Location Name": "AAPHC Northwest Family Medical and Dental Center (COVID Vaccine)", "County": "Dougherty", "Address": "2403-A Osler Court\nAlbany, GA 31707", "address-parts": {"address-line1": "2403-A Osler Court", "locality": "Albany", "administrative-area": "GA", "postal-code": "31707", "country": ""}, "phone-numbers": [{"href": "tel:+1-229-338-7589", "label": "COVID Vaccine Line:"}], "contact-links": [{"href": "https://www.aaphc.org", "label": "COVID-19 Vaccine Appointments"}], "node-id": "node/43421", "last-updated": "1611952801"}
{"Location Name": "Adams Drug Store", "County": "Crisp", "Address": "408 E 16th Ave\nSuite B\nCordele, GA 31015", "address-parts": {"address-line1": "408 E 16th Ave", "address-line2": "Suite B", "locality": "Cordele", "administrative-area": "GA", "postal-code": "31015", "country": ""}, "phone-numbers": [{"href": "tel:+1-229-273-3433", "label": "COVID Vaccine Line:"}], "contact-links": [], "node-id": "node/42341", "last-updated": "1610745772"}
{"Location Name": "Alamo - CareConnect Family Practice", "County": "Wheeler", "Address": "9 West Main Street\nAlamo, GA 30411", "address-parts": {"address-line1": "9 West Main Street", "locality": "Alamo", "administrative-area": "GA", "postal-code": "30411", "country": ""}, "phone-numbers": [{"href": "tel:+1-912-568-1438", "label": "COVID Vaccine Line:"}], "contact-links": [], "node-id": "node/42876", "last-updated": "1611237595"}
{"Location Name": "Allwell Pharmacy (COVID Vaccine)", "County": "Gwinnett", "Address": "1235 Indian Trail\nSuite B-100\nNorcross, GA 30093", "address-parts": {"address-line1": "1235 Indian Trail", "address-line2": "Suite B-100", "locality": "Norcross", "administrative-area": "GA", "postal-code": "30093", "country": ""}, "phone-numbers": [{"href": "tel:+1-770-858-5008", "label": "COVID Vaccine Line:"}], "contact-links": [{"href": "https://allwellpharmacy.us/", "label": "COVID-19 Vaccine Appointments"}], "node-id": "node/45241", "last-updated": "1618928861"}
{"Location Name": "Appling County Health Department (COVID Vaccine)", "County": "Appling", "Address": "283 Walnut Street\nBaxley, GA 31515", "address-parts": {"address-line1": "283 Walnut Street", "locality": "Baxley", "administrative-area": "GA", "postal-code": "31515", "country": ""}, "phone-numbers": [{"href": "tel:+1-855-473-4374", "label": "Primary:"}], "contact-links": [], "node-id": "node/38161", "last-updated": "1610746107"}
```

<details>
<summary>Formatted data sample</summary>

```json
[
  {
    "Location Name": "1st Care Management",
    "County": "Fulton",
    "Address": "1120 Hope Rd\nSuite 210\nSandy Springs, GA 30350",
    "address-parts": {
      "address-line1": "1120 Hope Rd",
      "address-line2": "Suite 210",
      "locality": "Sandy Springs",
      "administrative-area": "GA",
      "postal-code": "30350",
      "country": ""
    },
    "phone-numbers": [],
    "contact-links": [
      {
        "href": "https://1stcaremanagement.com/",
        "label": "COVID-19 Vaccine Appointments"
      }
    ],
    "node-id": "node/42871",
    "last-updated": "1611102629"
  },
  {
    "Location Name": "AAPHC Northwest Family Medical and Dental Center (COVID Vaccine)",
    "County": "Dougherty",
    "Address": "2403-A Osler Court\nAlbany, GA 31707",
    "address-parts": {
      "address-line1": "2403-A Osler Court",
      "locality": "Albany",
      "administrative-area": "GA",
      "postal-code": "31707",
      "country": ""
    },
    "phone-numbers": [
      { "href": "tel:+1-229-338-7589", "label": "COVID Vaccine Line:" }
    ],
    "contact-links": [
      {
        "href": "https://www.aaphc.org",
        "label": "COVID-19 Vaccine Appointments"
      }
    ],
    "node-id": "node/43421",
    "last-updated": "1611952801"
  },
  {
    "Location Name": "Adams Drug Store",
    "County": "Crisp",
    "Address": "408 E 16th Ave\nSuite B\nCordele, GA 31015",
    "address-parts": {
      "address-line1": "408 E 16th Ave",
      "address-line2": "Suite B",
      "locality": "Cordele",
      "administrative-area": "GA",
      "postal-code": "31015",
      "country": ""
    },
    "phone-numbers": [
      { "href": "tel:+1-229-273-3433", "label": "COVID Vaccine Line:" }
    ],
    "contact-links": [],
    "node-id": "node/42341",
    "last-updated": "1610745772"
  },
  {
    "Location Name": "Alamo - CareConnect Family Practice",
    "County": "Wheeler",
    "Address": "9 West Main Street\nAlamo, GA 30411",
    "address-parts": {
      "address-line1": "9 West Main Street",
      "locality": "Alamo",
      "administrative-area": "GA",
      "postal-code": "30411",
      "country": ""
    },
    "phone-numbers": [
      { "href": "tel:+1-912-568-1438", "label": "COVID Vaccine Line:" }
    ],
    "contact-links": [],
    "node-id": "node/42876",
    "last-updated": "1611237595"
  },
  {
    "Location Name": "Allwell Pharmacy (COVID Vaccine)",
    "County": "Gwinnett",
    "Address": "1235 Indian Trail\nSuite B-100\nNorcross, GA 30093",
    "address-parts": {
      "address-line1": "1235 Indian Trail",
      "address-line2": "Suite B-100",
      "locality": "Norcross",
      "administrative-area": "GA",
      "postal-code": "30093",
      "country": ""
    },
    "phone-numbers": [
      { "href": "tel:+1-770-858-5008", "label": "COVID Vaccine Line:" }
    ],
    "contact-links": [
      {
        "href": "https://allwellpharmacy.us/",
        "label": "COVID-19 Vaccine Appointments"
      }
    ],
    "node-id": "node/45241",
    "last-updated": "1618928861"
  },
  {
    "Location Name": "Appling County Health Department (COVID Vaccine)",
    "County": "Appling",
    "Address": "283 Walnut Street\nBaxley, GA 31515",
    "address-parts": {
      "address-line1": "283 Walnut Street",
      "locality": "Baxley",
      "administrative-area": "GA",
      "postal-code": "31515",
      "country": ""
    },
    "phone-numbers": [{ "href": "tel:+1-855-473-4374", "label": "Primary:" }],
    "contact-links": [],
    "node-id": "node/38161",
    "last-updated": "1610746107"
  }
]
```

</details>

## Before Opening a PR
- [x] I tested this using the CLI
   ```sh
   poetry run vaccine-feed-ingest fetch --state ga
   poetry run vaccine-feed-ingest parse --state ga
   ```
- [x] I ran auto-formatting: `poetry run tox -e lint-fix`

### CODE REVIEW

File: vaccine_feed_ingest/runners/ga/dph/parse.py [hunk 1]
```diff
@@ -2,29 +2,98 @@
...
     location_rows = doc.select("#datatable > tbody > tr")
+    locations = []
+    for row in location_rows:
+        cells = row.find_all("td")
+        location = {
+            # these first 3 items are for backwards compatibility with the
+            # previous parser iteration
+            "Location Name": cells[0].text.strip(),
+            "County": cells[1].text.strip(),
+            "Address": cells[2].text.strip(),
```
Comment (eliblock) [inline vaccine_feed_ingest/runners/ga/dph/parse.py:76]: Thank you for doing this.

File: vaccine_feed_ingest/runners/ga/dph/parse.py [hunk 2]
```diff
@@ -2,29 +2,98 @@
...
-    input_dir = pathlib.Path(sys.argv[2])
+def parse_location(input_file: pathlib.Path) -> dict:
+    result: dict = {"phone-numbers": [], "contact-links": []}
+    with input_file.open() as f:
+        doc = BeautifulSoup(f, "html.parser")
+    if doc is None:
+        raise Exception("failed to set up beautiful soup")
+    main_el = doc.find(id="main-content")
+    if main_el is None:
+        raise Exception("main content block has changed classes")
```
Comment (eliblock) [inline vaccine_feed_ingest/runners/ga/dph/parse.py:21]: Thank you for adding errors like these which will help us maintain this over time.

**Review by eliblock — COMMENTED** [review]
It looks like line 26 of the downloaded HTML files is often something like `<meta property="ga:citation:metadata" last_updated="1610746863" internal_url="node/37426" />`.

This looks to me like the update timestamp for the individual data fields, as well as a potential stable ID. What do you think?

**Comment by Graham42** [comment]
The internal_url `node/###` also aligns with `<link rel="shortlink" href="https://dph.georgia.gov/node/###" />`, makes sense to me that this is probably a stable ID. I can add it.

A quick sampling of the `last_updated` for different files also looks like it's a real last updated and not just "whenever you fetched the page" like some websites have 😆. I'll add this too.

**Comment by Graham42** [comment]
I've updated the data sample in the PR description with the new changes

**Review by eliblock — APPROVED** [review]
Thank you for adding that - and the additional fields overall!

As a heads up, @ericrobinsondev has an open PR for a normalizer for this data at #226 (I want to avoid duplicating work!). Hopefully they will be able to pick up on the additional parsed fields.

**Pull request merged**